Convert OpenMP parallelization to OneAPI::TBB

dbs4261 opened the pull request:
OpenMP acceleration has been migrated to use oneapi::TBB.

## Type

<!--- Select with 'x' and link to a related issue. What types of changes does your code introduce? -->

-   [ ] Bug fix (non-breaking change which fixes an issue): Fixes #
-   [x] New feature (non-breaking change which adds functionality). Resolves #
-   [x] Breaking change (fix or feature that would cause existing functionality to not work as expected) Resolves #N/A

## Motivation and Context

Many components of Open3D imply an eventual shift away from OpenMP to TBB. This includes some sections where tbb is only used on one platform as 2D loop unrolling isn't supported on Win32. Lastly, by using multiple parallelization paradigms, nested parallelism is problematic. When using some Open3D methods from a TBB context, an OpenMP thread pool is created for each TBB thread.

## Checklist:

<!--- Go over all the following points, and put an `x` in all the boxes that
apply.  If you're unsure about any of these, don't hesitate to ask. We're here
to help! -->

-   [x] I have run `python util/check_style.py --apply` to apply Open3D **code style**
    to my code.
-   [ ] This PR changes Open3D behavior or adds new functionality.
    -   [ ] Both C++ (Doxygen) and Python (Sphinx / Google style) **documentation** is
        updated accordingly.
    -   [x] I have added or updated C++ and / or Python **unit tests** OR included **test
        results** (e.g. screenshots or numbers) here.
-   [x] I will follow up and update the code if CI fails.
    <!-- In case I am unavailable later -->
-   [x] For fork PRs, I have selected **Allow edits from maintainers**.

## Description

Updated parallel for sections to use tbb::parallel_for. Adapted most loops that performed reductions with either omp reduction clauses or with critical sections to tbb::parallel_reduce implementations. Some of which required custom reduction objects instead of using lambdas. Added an atomic version of the ProgressBar for use with TBB.

There is still work to be done in documentation. This will break any user code that directly uses `ParallelForCPU` as OpenMP critical sections will no longer work. Additionally, TBB has no approach for setting the maximum number of threads like OpenMP does with `OMP_NUM_THREADS`. In C++ code a `tbb::global_control` object could be used, but it is unclear to how to provide that sort of functionality for python users.

dbs4261 committed:
Switched from using OpenMP for parallelism to using oneAPI::TBB (though not using the oneapi scope). Untested but building.

dbs4261 committed:
Swichted usage of std::mutex to tbb::mutex for consistency when used in conjunction with tbb parallel constructs.

dbs4261 committed:
Fixed bug in CPU reduction

dbs4261 committed:
Shift atomic to outside of RW mutex in PointCloudSegmentation.cpp

dbs4261 committed:
Switch from using a mutex to a concurrent vector for parallelization of self intersecting triangles.

dbs4261 committed:
Get maximum threads from TBB instead of OpenMP

dbs4261 committed:
Updates to ProgressBar.h/.cpp Including separating code for writing the progress bar into its own function and a bulk inplace add function operator+=. Also added TBBProgressBar. It does not inherit from ProgressBar as it uses an atomic for counting and has slightly different internals to use that atomicity.

dbs4261 committed:
Updated ClusterDBSCAN in PointCloudCluster.cpp to bulk update the progress bar to limit spinning on the mutex.

dbs4261 committed:
Applied Open3D style

dbs4261 commented:
Ok, I ran my tests in my development environment. Guess I should use the docker containers to replicate the CI environment and figure out those tests.

ssheorey commented:
Hi @dbs4261 thanks for picking this up!

Possibly fixes #6544

errissa commented:
@dbs4261 Thanks for working on this! I just tested this PR on my Mac and got numerous TBB related compilation errors. I tried using the Homebrew version of TBB as well as the "build from source" configuration. There appear to be functions that this PR uses that are missing from the Homebrew and "build from source" versions of TBB on Mac.

I know this PR is still draft but wanted to report what I had found. Please let me know if you need any help testing/diagnosing issues on Mac.

dbs4261 commented:
Hi @ssheorey this PR likely wont fix that issue as I haven't yet changed how the TBB dependency is being accessed. This is likely also why @errissa is facing issues building on Mac.

@errissa is homebrew pulling the OneAPI version of TBB? If you can provide me with the version of TBB you tried and the compiler errors I can take a look and figure out which version is required and work that into the PR.

ssheorey commented:
@dbs4261 yes, you are right about not fixing #6544. We should update to the latest oneTBB as part of this PR to fix that though.

This is the latest version of oneTBB and is available for all platforms on github: 

https://github.com/oneapi-src/oneTBB/releases/tag/v2021.11.0

The naming is off - this was released in Nov 2023.

I think this should also resolve @errissa 's issues on macOS.

dbs4261 commented:
I agree that setting the version requirement for TBB should be part of this PR. Based on the ubuntu failure in CI, its the `collaborative_call_once` header that is missing. The TBB repo says the header hasnt been modified in 3 years, so I would think that any version that reports 2021+ should be fine. What does Open3D CI currently use for TBB?

errissa commented:
@dbs4261 @ssheorey is correct about the oneTBB version. Homebrew's most recent version is 2021.11.0 so if this PR builds successfully against it, it would solve the MacOS issue I experienced.

ssheorey commented:
Hi @dbs4261 , our usual policy is to upgrade to the latest version available, but set minimum version to what is required to make everything work. This helps to "future-proof" the updated code as much as possible by incorporating the latest bugfixes. Official binaries will be built with the latest version, but also allows the library to build on older versions by users.

dbs4261 committed:
Marked single argument constructors for reductions as explicit to please codacy

dbs4261 committed:
Explicitly load atomics in calls to utility::Log*(...) because newer versions of format wont automatically convert it to its underlying type.

ssheorey commented on cpp/open3d/core/ParallelFor.h:
```diff
@@ -79,10 +81,13 @@ void ParallelForCPU_(const Device& device, int64_t n, const func_t& func) {
-        func(i);
-    }
+    tbb::parallel_for(tbb::blocked_range<int64_t>(0, n, 32),
+                      [&func](const tbb::blocked_range<int64_t>& range) {
```
Also, avoid using "magic numbers" (32). I think you have a `GetDefaultChunkSize()` function.

dbs4261 replied:
It will use up to the number of threads in task arena that called it. As for the chunk size, see [my other comment](https://github.com/isl-org/Open3D/pull/6626#discussion_r1505092754).

ssheorey reviewed (commented):
[Initial look]

dbs4261 commented on 3rdparty/mkl/tbb.cmake:
```diff
@@ -26,13 +26,10 @@ find_package(Git QUIET REQUIRED)
     PREFIX tbb
-    URL https://github.com/wjakob/tbb/archive/141b0e310e1fb552bdca887542c9c1a8544d6503.tar.gz # Sept 2020
-    URL_HASH SHA256=bb29b76eabf7549660e3dba2feb86ab501469432a15fb0bf2c21e24d6fbc4c72
+    URL https://github.com/oneapi-src/oneTBB/archive/refs/tags/v2021.4.0.tar.gz
```
No reason why not. I just put in the older version that had all the features I used.

dbs4261 commented on cpp/open3d/utility/Parallel.cpp:
```diff
@@ -7,51 +7,21 @@
-        return "";
-    }
-}
+int EstimateMaxThreads() { return tbb::this_task_arena::max_concurrency(); }
```
No, the number of tasks is determined by the caller. A caller could be using a small task arena to deal with IO, while a larger arena deals with processing something else. This actually brings up an issue that I don't yet know how to solve. TBB sets the maximum concurrency with a C++ variable that follows scope rules but doesn't need to be passed to functions. So I don't know how a python user would set the concurrency limit yet.

dbs4261 commented on cpp/open3d/utility/Parallel.cpp:
```diff
@@ -7,51 +7,21 @@
-    return 1;
-#endif
+std::size_t& DefaultGrainSizeTBB() noexcept {
+    static std::size_t GrainSize = 256;
```
Honestly, I was guessing at grain size [from this](https://oneapi-src.github.io/oneTBB/main/tbb_userguide/Controlling_Chunking_os.html), but it really should be picked based off of profiling. My understanding is that the grain size provides loose guidance to TBB's automatic chunking mechanism. It works similarly to `omp schedule(guided)`. Overall the goal is to provide plenty of work to each thread so the overhead of chunking is minimized, but small enough chunks that the scheduler can go back in a steal some if one of the threads gets held up. It might be worth taking another pass through the grain sizes that I put in and set them as a magic number times the `DefaultGrainSizeTBB` (which is mutable). That way the chunk size could be higher for doing a single operation with tensors, and smaller when looping through complex sections like in RANSAC.

ssheorey commented:
[Notes about linking and binary distribution]

For linking TBB, recommendation is to link dynamically. For C++ binaries and applications, we will distribute TBB DLL along with the Open3D DLL.
https://github.com/oneapi-src/oneTBB/issues/646

For Python, TBB libraries are available through PyPI, so we can add these as dependencies to `requirements.txt`
https://community.intel.com/t5/Intel-oneAPI-Threading-Building/How-to-ship-a-package-using-TBB-on-PyPI-manylinux/m-p/1227574

benjaminum commented on cpp/open3d/utility/ProgressBar.h:
```diff
@@ -15,30 +18,57 @@ namespace utility {
 public:
     ProgressBar(size_t expected_count,
-                const std::string &progress_info,
+                std::string progress_info,
```
Why has the const been removed here?

dbs4261 replied:
It has to be copied into the object, so it passed by value into the constructor and then by move into the member variable.

ssheorey committed:
style fix

dbs4261 committed:
Fixed incorrect include in Parallel.h

dbs4261 committed:
Updated ProgressBar.h to use size_t in std namespace for consistency

dbs4261 committed:
Switched std::lock_guard<std::mutex> for DiscreteGenerator to use tbb::spin_mutex::scoped_lock.

dbs4261 committed:
Switch to using tbb::parallel_reduce on all platforms.

dbs4261 committed:
Fixed progress bar bug that allow for multiple threads writing to the terminal.

dbs4261 committed:
Fixed issue with CPU reduction. Need to exit early if no work is to be done to prevent assignment to the output pointer.

dbs4261 committed:
Updated MemoryManagerStatistic to use a tbb::spin_mutex instead of a std::mutex

dbs4261 committed:
Updated point cloud segmentation to generate random samples using the global mutex from utilities::random.

dbs4261 committed:
Updated example usage of global mutex and engine access to reflect the TBB types.

dbs4261 committed:
Merge branch 'main' into omp2tbb

dbs4261 committed:
Turn TBB into publicly linked library

PKizzle commented:
Is there anything that can be done to fix the two failing tests?

```
_______________________ test_get_surface_area[device0] ________________________

device = CPU:0

    @pytest.mark.parametrize("device", list_devices())
    def test_get_surface_area(device):
        # Test with custom parameters.
        cube = o3d.t.geometry.TriangleMesh.create_box(float_dtype=o3c.float64,
                                                      int_dtype=o3c.int32,
                                                      device=device)
        np.testing.assert_equal(cube.get_surface_area(), 6)
    
        empty = o3d.t.geometry.TriangleMesh(device=device)
        empty.get_surface_area()
        np.testing.assert_equal(empty.get_surface_area(), 0)
    
        # test noncontiguous
        sphere = o3d.t.geometry.TriangleMesh.create_sphere(device=device)
        area1 = sphere.get_surface_area()
        sphere.vertex.positions = sphere.vertex.positions.T().contiguous().T()
        sphere.triangle.indices = sphere.triangle.indices.T().contiguous().T()
        area2 = sphere.get_surface_area()
>       np.testing.assert_almost_equal(area1, area2)

python\test\t\geometry\test_trianglemesh.py:859: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

args = (12.501888275146484, 12.501884460449[219](https://github.com/isl-org/Open3D/actions/runs/11150773935/job/30997043371#step:6:220)), kwds = {}

    @wraps(func)
    def inner(*args, **kwds):
        with self._recreate_cm():
>           return func(*args, **kwds)
E           AssertionError: 
E           Arrays are not almost equal to 7 decimals
E            ACTUAL: 12.501888275146484
E            DESIRED: 12.501884460449219

C:\hostedtoolcache\windows\Python\3.11.9\x64\Lib\contextlib.py:81: AssertionError
```

dbs4261 commented:
I can take a look, they are both in the python library, correct?

PKizzle commented:
Yes, I guess so. You can find them here:

- python\test\t\geometry\test_trianglemesh.py:859
- python\test\test_color_map_optimization.py:49

ssheorey committed:
Merge branch 'main' into omp2tbb; remove remaining OpenMP dependency

Resolve conflicts from the origin/main update, keeping the PR's intent
(oneAPI TBB replaces all OpenMP parallelism) while preserving the features
main added in the meantime:

- ReductionCPU: keep main's two-pass / parallel-dim split for arg-reduction,
  converted to TBB. Use parallel_deterministic_reduce for all floating point
  reductions so results no longer depend on thread count (this fixes the
  test_get_surface_area and test_color_map failures reported on the PR).
- registration/Feature: keep main's SPFH/FPFH subset-index filtering.
- PointCloudSegmentation: keep main's pre-generated RANSAC samples.
- ColorMapUtils: keep main's utility::optional -> std::optional migration.
- VoxelBlockGridImpl: keep main's SYCL atomic min/max paths.

Complete the OpenMP removal so Open3D no longer has an OpenMP dependency:

- Convert the last OpenMP loops (Smoothing.h, PointCloudSmoothing.cpp,
  registration/Feature.cpp) to tbb::parallel_for.
- Remove the WITH_OPENMP option, the OpenMP find_package, the libomp
  packaging for macOS wheels, and the WITH_OPENMP build-config entry.
- Remove utility::OMPProgressBar (superseded by TBBProgressBar) and
  utility::GetThreadNum() / InParallel(), which are OpenMP-only.
- SurfaceReconstructionPoisson: always use PoissonRecon's std::thread pool.
- Replace examples/cpp/OpenMP.cpp with examples/cpp/Parallelism.cpp, which
  demonstrates parallel_for, parallel_reduce and task_arena.

Add a replacement for OMP_NUM_THREADS, which TBB has no equivalent of:
utility::SetMaxThreads() (tbb::global_control), exposed to Python as
open3d.utility.set_max_threads() / get_max_threads(). The reconstruction
system examples use it instead of setting OMP_NUM_THREADS.

Fix bugs found while resolving and reviewing the conversion:

- Registration.cpp: atomic_min's compare-exchange loop exited on contention
  instead of retrying, so the RANSAC iteration bound could be lost.
- Registration.cpp: RANSACCorrespondenceReduction's splitting constructor did
  not size ransac_corres, so split bodies wrote out of bounds.
- LegacyReconstructionUtil.h: PointCloudIntegrator never accumulated into
  fragment, so IntegrateFragmentRGBD produced an empty fragment.
- UniformTSDFVolume::ExtractVoxelGrid wrote into a std::unordered_map from a
  parallel_for; collect into a concurrent_vector and insert serially.
- ColorMapUtils: inverted empty check when averaging invisible-vertex colors.
- ProgressBar::operator+=(n) advanced by 1 instead of n.
- EstimateMaxThreads() ignored tbb::global_control limits.
- Fix '#defined' typos that would break CUDA/SYCL builds of
  VoxelBlockGridImpl.h and FillInLinearSystemImpl.h.
- Consolidate two copies of a double atomic-add helper (whose feature-test
  guard was malformed) into utility::AtomicAdd.
- VoxelBlockGridImpl.h: correct an off-by-one in the ExtractTriangleMesh
  assertion added in #7360; 0 is a valid linear index, and the always-on
  device assert made CUDA tests flaky.

Require oneTBB >= 2021.4.0 (for tbb/collaborative_call_once.h) and link TBB
as a public dependency, since TBB types appear in public Open3D headers.

ssheorey committed:
Fix TBB parallelism bugs found in performance review

Performance review of the OpenMP->TBB conversion (PR #6626) surfaced two
real scaling bugs and three formatter-injected @AGENT review comments
to address:

- ParallelFor.h: the ISPC-vectorized CPU path pre-splits work into
  num_threads chunks, then called ParallelForCPU_ with the default
  grain size (32). Since num_threads (e.g. 20) < 32, tbb::parallel_for
  never split the range and ran it on a single thread (measured 1.00x
  speedup regardless of thread count). Added an explicit grain_size
  parameter to ParallelForCPU_ and pass grain_size=1 for this
  pre-chunked path. Recovers 2.4-4.8x speedup for sqrt/exp/abs and
  other unary ops.

- Parallel.cpp: DefaultGrainSizeTBB() was 256, which capped small
  inputs (N~1000) to ceil(N/256)=4 chunks regardless of thread count
  (measured ~3.7x max speedup at N=1000 for estimate_normals-style
  workloads). Empirical sweep across thread counts and input sizes
  shows 64 matches 256 for large inputs and is up to ~2x faster for
  inputs of a few thousand items. DefaultGrainSizeTBB2D() (32, used
  for cheap element-wise kernels) is unchanged.

- Registration.cpp: atomic_min now uses explicit
  std::memory_order_relaxed for both load and compare_exchange_weak,
  since min_val is only an optimization hint (RANSAC iteration bound)
  with no other data published through it.

- ProgressBar.h: documented that ProgressBar is not thread-safe
  (races on current_count_ and terminal output), contrasted with
  TBBProgressBar which is safe for use inside a parallel region.

- UniformTSDFVolume.cpp: ExtractVoxelGrid() replaced
  tbb::concurrent_vector with tbb::enumerable_thread_specific<vector>,
  collecting per-thread results and inserting into the (non-concurrent)
  voxel map after the parallel loop. Avoids concurrent_vector's
  per-element synchronization and segmented layout; measured
  ~1.2-1.6x faster.

Verified: full C++ test suite (1825 passed) and Python test suite
(965 passed, 13 skipped) both pass after each change.

ssheorey committed:
Finalize TBB cleanup and CI packaging fixes

ssheorey committed:
Merge remote-tracking branch 'origin/main' into omp2tbb

ssheorey committed:
Fix TBB review findings

ssheorey committed:
Merge branch 'main' into omp2tbb

ssheorey committed:
Fix CI failures: SYCL mutex capture and CMake TBB export

- VoxelBlockGridImpl.h: Exclude tbb::spin_mutex capture for SYCL builds
  using BUILD_SYCL_MODULE guard. SYCL device code cannot handle TBB
  mutex types as kernel parameters; the SYCL path already uses device
  atomics inside the lambda instead.

- find_dependencies.cmake: Add PUBLIC flag to oneAPI TBB find_package
  call. When BUILD_SHARED_LIBS=ON (e.g., Windows XPU builds), the
  target must be in the Open3DTargets export set. Without PUBLIC,
  the install(EXPORT) step fails because Open3D links against
  Open3D::3rdparty_tbb but the target is not in any export set.

ssheorey committed:
Merge branch 'omp2tbb' of https://github.com/dbs4261/open3d into omp2tbb

ssheorey committed:
Fix SYCL device preprocessor check in VoxelBlockGridImpl

ssheorey committed:
correct #defines for each of cpu, cuda, sycl

ssheorey committed:
ubuntu-sycl.yml: fix YAML/ccache bugs, run CPU-only SYCL tests via docker_test.sh

- Fix broken YAML indentation that made the workflow file invalid.
- Fix ccache upload referencing .tar.gz when Dockerfile.ci/docker_build.sh
  always produce .tar.xz.
- Split docker_test.sh's cpp_python_linking_uninstall_test() into
  docker_run_setup()/cpp_test()/python_test()/linking_test()/uninstall_test(),
  keeping the combined function as a thin wrapper for existing callers
  (openblas/cpu/cuda configs unaffected).
- Guard --device=/dev/dri behind an existence check so SYCL tests work on
  GitHub-hosted CPU-only runners (no Intel GPU) as well as on GCE VMs with
  a real Intel GPU; the SYCL runtime already falls back to the CPU device
  automatically when no GPU is present.
- sycl-shared/sycl-static docker_test.sh targets accept an optional
  cpp|python phase argument.
- build-lib job now runs only the C++ phase via docker_test.sh; Python
  testing remains solely in build-wheel (against the real per-version
  wheel), avoiding duplicate test runs.
